### ISSUE TITLE
Unexpected behavior on model_dump when replacing type

### ISSUE THREAD

**Issue by noddycode** [pydantic/pydantic#6916]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

This actually looks like it could be related to a closed issue, #5784, but I'm using a later version of Pydantic than mentioned in that issue and am still facing a similar problem. When replacing a list field with a list containing another type, `model.model_dump` outputs a tuple containing the expected output along with a `None` value. It is somehow related to `extras='allow'` on the original (unused) model.

I tried to play around and minimize the example as much as possible, but I'm still not sure of the actual source of this behavior.

### Example Code

```Python
from typing import Optional, List
from pydantic import BaseModel, ConfigDict


class MyUnusedModel(BaseModel):
    model_config = ConfigDict(extra="allow")


class MyNestedModel(BaseModel):
    int_field: int = 1


class MyModel(BaseModel):
    model_config = ConfigDict(extra='allow')

    list_field: Optional[List[MyUnusedModel]] = None


model = MyModel()
model.list_field = [MyNestedModel()]

print(model.model_dump())
# {'list_field': [({'int_field': 1}, None)]}
# stderr: UserWarning: Pydantic serializer warnings: Expected `general-fields` but got `tuple` - serialized value may not be as expected
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.1.1
pydantic-core version: 2.4.0
pydantic-core build: profile=release pgo=false mimalloc=true
install path: <LOCAL_VENV>
python version: 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)]
platform: Windows-10-10.0.19044-SP0
optional deps. installed: ['email-validator', 'typing-extensions']
```

**Comment by dmontagu** [comment]
It's not surprising that this has problems because `MyNestedModel` is not an instance of `MyUnusedModel`, so you get this error when you do a type-invalid assignment like you have.

If you change it to:
```python
from typing import Optional, List
from pydantic import BaseModel, ConfigDict


class MyUnusedModel(BaseModel):
    model_config = ConfigDict(extra="allow")


class MyNestedModel(MyUnusedModel):
    int_field: int = 1


class MyModel(BaseModel):
    model_config = ConfigDict(extra='allow')

    list_field: Optional[List[MyUnusedModel]] = None


model = MyModel()
model.list_field = [MyNestedModel()]

print(model.model_dump())
#> {'list_field': [{}]}
```
the error goes away though.

If you want to serialize the fields from the list items, you can use `SerializeAsAny`, which coincidentally resolves the issue whether `MyNestedModel` is a subclass of `MyUnusedModel` or not:
```python
from typing import Optional, List
from pydantic import BaseModel, ConfigDict, SerializeAsAny


class MyUnusedModel(BaseModel):
    model_config = ConfigDict(extra="allow")


class MyNestedModel(BaseModel):
    int_field: int = 1


class MyModel(BaseModel):
    model_config = ConfigDict(extra='allow')

    list_field: Optional[List[SerializeAsAny[MyUnusedModel]]] = None


model = MyModel()
model.list_field = [MyNestedModel()]

print(model.model_dump())
#> {'list_field': [{'int_field': 1}]}
```

That said, I think the current behavior in your test script above is definitely not ideal — @davidhewitt or @adriangb I think you guys might be able to get to the bottom of this quickly, or at least quicker tahn me, any chance one of you could take a look? Specifically, the fact that a non-subclass model instance is getting converted into a tuple when it probably should be just the dict of field values (and extra I guess).

**Comment by noddycode** [comment]
Yeah, I'm more concerned about it being a breaking change than something we really *should* be doing.

Our use case is a little complicated since we're working with models generated by a... messy schema, so a lot of these kinds of issues pop up due to various ways we've tried to simplify working with those generated models.

**Comment by dmontagu** [comment]
Well, for what it's worth I do believe there is a bug here that should (and can) be fixed.

But you'll probably still get warnings when serializing like this unless you use the `SerializeAsAny` annotation.

**Comment by alexgaspard** [comment]
I've got a similar issue 

```python
from typing import Generic, TypeVar

from pydantic import BaseModel

T = TypeVar("T", bound=BaseModel)


class PaginatedResult(BaseModel, Generic[T]):
    list: list[T]
    total: int


class People(BaseModel):
    name: str


def get_people() -> PaginatedResult[People]:
    return PaginatedResult(list=[People(name="alice")], total=2)


print(get_people())
print(get_people().model_dump())
```

Result:
```shell
list=[People(name='alice')] total=2
{'list': [{}], 'total': 2}}
```

`model_dump` works correctly if I use 

```python
def get_people() -> PaginatedResult[People]:
    return PaginatedResult[People](list=[People(name="alice")], total=2)
```

but pylint complains that `PaginatedResult[People] is not callable`.  
It also works with the workaround 

```python
T = TypeVar("T", bound=SerializeAsAny[BaseModel])
```
(with pydantic 2.1.1 and pydantic_core 2.4.0)

**Comment by davidhewitt** [comment]
From some investigation, the cause of the original strange-looking warning is a combination of the fact that `MyUnusedModel` allows extra and `MyNestedModel` doesn't, and the fact that the types have been mixed up.

In the model serializer we just look for `__dict__` when serializing, so the `MyNestedModel` instance is allowed through the `MyUnusedModel` serializer

https://github.com/pydantic/pydantic-core/blob/87b4789a1943514785c967aca2a059e81fd2c2f1/src/serializers/type_serializers/model.rs#L117

... so the combination of `__dict__` and `__pydantic_extra__` is `({'int_field': 1}, None)`

This then doesn't match up with the type `Tuple[dict, dict]` which is expected when `extra='allow'` for the fields serializer.

https://github.com/pydantic/pydantic-core/blob/87b4789a1943514785c967aca2a059e81fd2c2f1/src/serializers/fields.rs#L121C43-L121C43

Probably the right fix is to make the fields serializer more lax to allow this to go through (and serialize properly), rather than to make the model serializer stricter.
